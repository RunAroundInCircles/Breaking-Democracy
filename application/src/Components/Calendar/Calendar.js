//Base code was found here: https://blog.flowandform.agency/create-a-custom-calendar-in-react-3df1bfd0b728
// The code needed to be recreated to work with the new version of datefns
import React from "react";
import {format,startOfWeek,addMonths,startOfMonth,addDays,subMonths,endOfWeek,endOfMonth,isSameMonth,isSameDay,isEqual, isAfter, isBefore, add}  from "date-fns";

/**
 * Calendar component of the app that allows the user to click different events
 * during the political campaign to influence candidates.
 * @extends React
 */
class Calendar extends React.Component {
  state = {
    currentMonth: new Date(2020, 2, 1, 0, 0, 0, 0), //The Current Month shown to the user
    selectedDate: new Date(2020, 2, 1, 0, 0, 0, 0), //The date of the clicked cell
    currentProgress: new Date(2020, 2, 1, 0, 0, 0, 0)
  };

  /**
   * Goes through all the dates and searches for an event that is on the same day as the date passed in.
   * This also goes through all the events already completed by the player and will remove them from the calendar
   * @param  {[Date]} date The date to search for.
   *  * @return {String} Returns an event if one exists, else it returns null.
   */
  containsDate(date) {
    var i;
    for(i = 0; i < this.props.events.length; i++) {

      if(isEqual(date, this.props.events[i].date)) {
        /*If any of the items in eventsCompleted (Found in MainPage are equal to the index then we do not render the item) */
        if (this.props.eventsCompleted.find(element => element.eventID == i)){
          return null;
        }
        else{
          return this.props.events[i].message;
        }
      }
    }
    return null;
  }

  /**
   * Renders the Month, year, and arrows to change months at the top of the calendar
   * @return {String} Creates the header for the calendar
   */
  renderHeader() {
    const dateFormat = "MMMM yyyy";

    return (
      <div className="header row flex-middle">
        <div className="col col-start">
        {/*Creates an arrow to allow the user to click to the previous month*/}
          <div className="icon" onClick={this.prevMonth}>
            chevron_left
          </div>
        </div>

        {/*Displays the month and year*/}
        <div className="col col-center">
          <span>{format(this.state.currentMonth, dateFormat)}</span>
        </div>
        {/*Creates an arrow to allow the user to click to the next month*/}
        <div className="col col-end" onClick={this.nextMonth}>
          <div className="icon">chevron_right</div>
        </div>
      </div>
    );
  }

  /**
   * Gives the date format of the days in the week list above the cells
   * @return {String} Returns the days of the week to be displayed in the calendar
   */
  renderDays() {
    const dateFormat = "eee"; //Gives the day as a 3 letter accronym
    const days = [];

    //Finds the start of the week based off the current month
    let startDate =  startOfWeek(this.state.currentMonth);

    //Adds each day to the week list
    for (let i = 0; i < 7; i++) {
      days.push(
        <div className="col col-center" key={i}>
          {format( addDays(startDate, i), dateFormat)}
        </div>
      );
    }

    return <div className="days row">{days}</div>;
  }

  /**
   * Renders each cell in the calendar and adds the day number
   * @return {div} Returns a div that represent a day as a cell in the calendar
   */
  renderCells() {



    const {currentMonth, selectedDate} = this.state;
    const monthStart =  startOfMonth(currentMonth);
    const monthEnd =  endOfMonth(monthStart);
    const startDate =  startOfWeek(monthStart);

    let day = startDate;
    console.log(day);
    const endDate =  endOfWeek(monthEnd);
    var gameRoundEnd = addDays(startOfWeek(this.state.currentProgress), 13);

    const dateFormat = "d";
    const rows = [];

    let days = [];

    let formattedDate = "";

    //Goes through every day in the month and formats them (Gives the correct day number to the cell)
    while (day <= endDate) {
      for (let i = 0; i < 7; i++) {
        formattedDate =  format(day, dateFormat);
        const cloneDay = day;
        var status = "enabled";

          if(!isSameMonth(day,startOfWeek(this.state.currentProgress))){
            status = "disabled";
          }

          if(isBefore(day, startOfWeek(this.state.currentProgress)) || isAfter(day, gameRoundEnd)){
            status = "disabled";
          }
        days.push(
          <div
            className={`col cell ${
            status}`}
            key={day}
          >
            <span className="number">{formattedDate}</span>
            <span className="bg">{formattedDate}</span>
            {this.containsDate(day)} {/*Gets the event for the day if it exists*/}
          </div>
        );
        day = addDays(day, 1);
      }
      rows.push(
        <div className="row" key={day}>
          {days}
        </div>
      );
      days = [];
    }

    return <div className="body">{rows}</div>;
  }

/**
 * When the right arrow is pressed by the user the month will increase by one
 * @return {int} Increments the month and the page will update to the next month
 */
  nextMonth = () => {
    this.setState({
      currentMonth: addMonths(this.state.currentMonth, 1)
    });
  };

/**
 * When the user clicks the left arrow the month will decrease by one
 * @return {int} Decrements the month and the page will update to the previous month
 */
  prevMonth = () => {
    this.setState({
      currentMonth: subMonths(this.state.currentMonth, 1)
    });
  };

/**
 * Renders all of the calendar
 * @return {div} Returns a div that represents the calendar
 */
  render() {
    return (
      <div className="calendar">
        {this.renderHeader()}
        {this.renderDays()}
        {this.renderCells()}
      </div>
    );
  }
}

export default Calendar;
