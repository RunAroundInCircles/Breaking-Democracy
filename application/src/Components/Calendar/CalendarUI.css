/*MIT License

Copyright (c) 2019 Caleb Logan

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


Base code was found here: https://blog.flowandform.agency/create-a-custom-calendar-in-react-3df1bfd0b728
We have changed some of the CSS to use different types of buttons
*/

@import url(https://fonts.googleapis.com/css?family=Open+Sans:300,400,700);
@import url(https://fonts.googleapis.com/icon?family=Material+Icons);

.icon {
  font-family: 'Material Icons', serif;
  font-style: normal;
  display: inline-block;
  vertical-align: middle;
  line-height: 1;
  text-transform: none;
  letter-spacing: normal;
  word-wrap: normal;
  white-space: nowrap;
  direction: ltr;

  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
  -moz-osx-font-smoothing: grayscale;
  font-feature-settings: 'liga';
}


/* VARIABLES */

:root {
  --main-color: #1a8fff;
  --completed-color: #2E8B57;
  --text-color: #777;
  --text-color-light: #ccc;
  --border-color: #eee;
  --bg-color: #f9f9f9;
  --neutral-color: #fff;
  --failed-color: #DC143C;
  --locked-color: #778899;
}


/* GENERAL */

* {
  box-sizing: border-box;

}

body {
    font-family: 'Open Sans', 'Helvetica Neue', 'Helvetica', 'Arial', sans-serif;
    font-size: 1em;
    font-weight: 300;
    line-height: 1.5;
    color: var(--text-color);
    background: var(--bg-color);
    height: fit-content;

}

header {
  display: block;
  width: 100%;
  padding: 1.75em 0;
  border-bottom: 1px solid var(--border-color);
  background: var(--neutral-color);
}

main {
  display: block;
  margin: 0 auto;
  margin-top: 5em;
  max-width: 50em;
}

/* GRID */

.row {
  margin: 0;
  padding: 0;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  width: 100%;
}

.row-middle {
  align-items: center;
}

.col {
  flex-grow: 1;
  flex-basis: 0;
  max-width: 100%;
}

.col-start {
  justify-content: flex-start;
  text-align: left;
}

.col-center {
  justify-content: center;
  text-align: center;
}

.col-end {
  justify-content: flex-end;
  text-align: right;
}


/* Calendar */

.calendar {
  display: block;
  position: relative;
  width: 100%;
  background: var(--neutral-color);
  border: 1px solid var(--border-color);
}

.calendar .header {
  text-transform: uppercase;
  font-weight: 700;
  font-size: 115%;
  padding: 1.5em 0;
  border-bottom: 1px solid var(--border-color);
}

.calendar .header .icon {
  cursor: pointer;
  transition: .15s ease-out;
}

.calendar .header .icon:hover {
  transform: scale(1.75);
  transition: .25s ease-out;
  color: var(--main-color);
}

.calendar .header .icon:first-of-type {
  margin-left: 1em;
}

.calendar .header .icon:last-of-type {
  margin-right: 1em;
}

.calendar .days {
  text-transform: uppercase;
  font-weight: 400;
  color: var(--text-color-light);
  font-size: 70%;
  padding: .75em 0;
  border-bottom: 1px solid var(--border-color);
}


.calendar .body .cell {
  position: relative;
  height: 7.4em;
  border-right: 1px solid var(--border-color);
  overflow: hidden;
  cursor: pointer;
  background: var(--neutral-color);
  transition: 0.25s ease-out;
}

.calendar .body .cell:hover {
  transition: 0.5s ease-out;
}

.calendar .body .selected {
  border-left: 10px solid transparent;
  border-image: linear-gradient(45deg, #1a8fff 0%,#53cbf1 40%);
  border-image-slice: 1;
}

.calendar .body .row {
  border-bottom: 1px solid var(--border-color);

}

.calendar .body .row:last-child {
  border-bottom: none;
}

.calendar .body .cell:last-child {
  border-right: none;
}

.calendar .body .cell .number {
  position: absolute;
  font-size: 82.5%;
  line-height: 1;
  top: .75em;
  right: .75em;
  font-weight: 700;
}

.calendar .body .disabled {
  color: var(--text-color-light);
  pointer-events: none;
  background: var(--bg-color);
}

.calendar .body .cell .bg {
  font-weight: 700;
  line-height: 1;
  color: var(--main-color);
  opacity: 0;
  font-size: 8em;
  position: absolute;
  top: -.2em;
  right: -.05em;
  transition: .25s ease-out;
  letter-spacing: -.07em;
}

.calendar .body .cell:hover .bg, .calendar .body .selected .bg  {
  opacity: 0.05;
  transition: .5s ease-in;
}

.calendar .body .col {
  flex-grow: 0;
  flex-basis: calc(100%/7.0005);
  width: calc(100%/7.0005);
}

.calendar-event-button {
  background-color: var(--main-color);
  box-sizing: border-box;
  width: 50%;
	border-style: solid;
	border-color: var(--border-color);
	color: white;
	padding: 5px 5px;
	text-align: center;
	text-decoration: none;
	display: inline-block;
	font-size: 16px;
}

.bg {
  pointer-events: none;
}
