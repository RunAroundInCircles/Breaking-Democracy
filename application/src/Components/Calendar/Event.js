import React,{Component} from 'react';
import {Link} from "react-router-dom";
import {Button} from 'react-bootstrap';

/**
 * Represents an event in the calendar. This is where users will be able
 * to start minigames and influence candidates
 * @extends React,react-router-dom
 */
class Event extends Component {
/**
 * Renders the alert and allows the user to go back to the original calendar
 * @return {Link} An event that is able to go back to the original calendar
 */
    render() {

        let route = "/Calendar/" + this.props.id;
        return (
            <Link to={route} className="calendar-event-link">
                <Button className="calendar-event-button">
                    {this.props.message}
				</Button>
            </Link>
        );
    }
}

export default Event;
