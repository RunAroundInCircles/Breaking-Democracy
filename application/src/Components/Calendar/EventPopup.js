import React from 'react';
import {Link} from "react-router-dom";
import {Button} from 'react-bootstrap';

/**
 * Creates a popup when an event is clicked on the calendar
 * @param {Properties} props Parameters needed to create the event, contains event which had an id, year, month, day, and a message
 * @return {div} Returns a div that blocks all click events below it and contains the popup window
 */
function EventPopup(props) {
    //The date of the event to display
    let date = new Date(props.event.year, props.event.month, props.event.day);

    //options is used to format how the date of the event is displayed
    let options = {year: 'numeric', month: 'short', day: 'numeric'};
    
    return (
        //This div covers the screen with an black opaque layer
         <div style={{display: 'flex',
            alignItems: 'center',
            justifyContent: 'center',
            position: 'fixed',
            width: '100vw',
            height: '100vh',
            zIndex: 1000,
            top: 0,
            left: 0,
            backgroundColor: 'rgba(0, 0, 0, .7)'
        }}>
            {/*This div is the body of the popup window containing the back button and the event info*/}
            <div style={{backgroundColor: 'rgba(255, 255, 255, 1)',
                borderRadius: '5px',
                width: '50vw',
                height: '50vh',
                position: 'absolute'
            }}>
                {/*This link acts as a back button allowing the user to redirect to Calendar*/}
                <Link to='/Calendar' >
                    <Button style={{top: 5, right: 5, position: 'absolute'}}>
						<span>X</span>
					</Button>
                </Link>
                {/*This div contains all of the information for the event*/}
                <div style={{justifyContent: 'center'}}>
                    <h1>{date.toLocaleDateString("en-US", options)}</h1>
                    <h2>{props.event.message}</h2>
					<Link>
						<Button>Choice 1</Button>
					</Link>
                </div>
            </div>
        </div>
    );
}

export default EventPopup;