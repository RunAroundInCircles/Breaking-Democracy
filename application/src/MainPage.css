.screen {
  position: relative;
}

.desktop {
  position: absolute;
  margin: 0;
  padding: 0;
  top: 0;
  left: 0;
  width: 100%;
  height: 100vh;
  z-index: -1;
}

Button {
  position:relative;
  z-index: 1;
}
