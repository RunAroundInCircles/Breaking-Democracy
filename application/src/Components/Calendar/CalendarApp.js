/*MIT License

Copyright (c) 2019 Caleb Logan

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

import React,{Component} from 'react';
import Calendar from './Calendar.js';
import './CalendarUI.css';
import Event from "./Event.js";


/**
 * Used to render the Calendar object
 * @extends React, Calendar, Event, ./CalendarUI.css
 */
class CalendarApp extends Component {
  /**
   * Converts the events passed in to objects containing a date and the message representing an event
   * @param {Property} props The parameters needed to setup the calendar
   */
    constructor(props) {
      super(props);

      this.state = {
          events: props.events.map((event) => {
            let date = new Date(event.year, event.month, event.day);
            return(
              {date: date, message: <Event key={event.id} message={event.message} date={date} id={event.id} status={event.status}/>}
            );
          })
      }
    }

	/**
   * Renders the calendar for the current month with all current events.
   * @return {div} Returns the calendar with all the events
   */
    render() {
        return(
            <div className="calendar-app">
                <Calendar callbackFromMain={this.callback} events={this.state.events} eventsCompleted={this.props.eventsCompleted} turnStartDate={this.props.turnStartDate}/>
            </div>
        )
    }

}

export default CalendarApp;
