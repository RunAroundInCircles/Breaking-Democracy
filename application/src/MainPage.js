/*MIT License

Copyright (c) 2019 Caleb Logan

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

import React,{Component} from 'react';
import './App.css';
import MapApp from './Components/Map/MapApp.js';
import MapRegion from './Components/Map/MapRegion.js';
import EmailApp from './Components/Email/EmailApp.js';
import CalendarApp from './Components/Calendar/CalendarApp.js';
import EchoApp from './Components/Echo/EchoApp.js'
import events from './Components/Calendar/EventList.json';
import emails from './Components/Email/EmailList.json';
import echos from './Components/Echo/echo.json';
import {Button} from 'react-bootstrap';
import EventPopup from './Components/Calendar/EventPopup.js';
import TimelineApp from './Components/Timeline/TimelineApp.js'
import './MainPage.css';
import desktop from './Resources/Title_Computer.png';
import Situations from './Components/Calendar/Situations.json';
import mainMusicMP3 from './Resources/Music/ThemeLoopable.mp3';
import mainMusicWAV from './Resources/Music/ThemeLoopable.wav';
import Intro from './Intro';
import GoodEnding from './GoodEnding';
import BadEnding from './BadEnding';
import {
  BrowserRouter as Router,
  Switch,
  Route,
  Link
} from "react-router-dom";
import { add, isBefore, isAfter, addDays } from 'date-fns';
import { createBrowserHistory } from "history";

/**
 * MainPage component of the app that renders and returns all the buttons
 * and allows you to switch between each page using reactRouter.
 * @extends React
 */
class MainPage extends Component{

	constructor(props) {
		super(props);

		this.state = {
			//Only stores red data to reduce unneccesary data storage
			pollData: {
				0: [75, 75, 75, 75, 75, 75, 75],
				1: [25, 25, 25, 25],
				2: [62, 33, 51, 83],
				3: [75, 50, 57],
				4: [38, 51],
				5: [18],
				6: [70, 25, 89, 34],
				7: [21, 12, 37]
			},
			//regionDistrictNames stores all of the names of the regions and districts to be displayed on the map
			//the first name in the array is the region, all subsequent are districts
			regionDistrictNames: {
					0: ["Saika","Rakka","Feidler","Larch","Broon","Lona La","Oglad","Prock"],
					1: ["Kaika","Ash","Holly","Kefler","Darby"],
					2: ["Flaze","Gretroit","Hearth","Magdo","Garde"],
					3: ["Libdove","Moka","Agon","Veera"],
					4: ["Osco","Proe","Haley"],
					5: ["Warren Central", "Warren Central"],
					6: ["Dukaste","Locke","Rehlat","Selia","Dukaste City"],
					7: ["Wegruesoe","Zaftan","Blektan","Wegruesoe City"]
			},
			//eventsCompleted is an array to hold all of the events that have been finished by the player after they complete them.
			eventsCompleted: [],
			currentEmails: [], //The current list of emails for the sprint we are on
			currentSprint: 1, //The current two week interval we are on
			//turnStartDate is the beginning Date for the game February 1, 2020, indicates the start of the turn in Calendar
			turnStartDate: new Date(2020, 2, 1, 0, 0, 0, 0),
			renderVideo: true, //Determines whether the intro video or the game should be rendered
			gameEnded: false, //Determines whether the user has finished the game
			lastEventID: 0, //The last event in the game
			hasPlayerWon: false, //Check to see if the player won
			playerScore: 0 //Holds the score of the player
		}

		this.callback = this.callback.bind(this);
		this.setCurrentEmail = this.setCurrentEmail.bind(this);
		this.ifExists = this.ifExists.bind(this);
		this.handleVideoEnd = this.handleVideoEnd.bind(this);
    	this.checkIfPlayerWon = this.checkIfPlayerWon.bind(this);

		var tempDate =  new Date(1 , 1, 1, 0, 0, 0, 0);
		//Find the eventID with the last day and month
		Object.values(events).map((event) => {
			let eventDate = new Date(event.year, event.month, event.day, 0, 0, 0, 0);
			if(isAfter(eventDate,tempDate)) {
					this.state.lastEventID = event.id;
					tempDate = eventDate;
			}
		});

		//Creates a history for the Router so that we can add './Email' to it
		//This allows us to skip the '/' page and go directly to './Email' instead
		//let createdHistory = createBrowserHistory();
		//createdHistory.push('./Email');
	}

    /**
  	 * Allows an external component to add entries to eventsCompleted and update the pollData
  	 * @param  {eventid}   eventsCompleted The id of the event completed.
  	 * @param  {percent}   eventsCompleted The percentage amount of change for the region's district
  	 */
  	callback = (eventid, percent) => {
  		//Get the region and district to change
  		var region = Math.floor(Math.random() * 8);
  		var district = Math.floor(Math.random() * this.state.pollData[region].length);

  		//Create the new completed event to add
  		var eventCompleted = {
  			eventID: eventid,
  			percent: percent,
  			region: this.state.regionDistrictNames[region][0], //Adds the name of the region
  			district: this.state.regionDistrictNames[region][district + 1] //Adds the name of the district
  		}

  		//Update the poll data
  		let updatedData = this.state.pollData;
  		updatedData[region][district] += (updatedData[region][district] * percent)

  		//Check that the updated poll data isn't over 100%
  		if(updatedData[region][district] > 100) {
  			updatedData[region][district] = 100;
  		}

		//Checks to see if the user has finished all events
		let updatedGameEnded = this.state.gameEnded;
		if(Object.values(this.state.eventsCompleted).length + 1 == Object.values(events).length) {
			updatedGameEnded = true;
		}

		//Get the event IDs between the two dates that need to be completed before the round can advance
		let eventsToComplete = this.getEventIDsBetween(this.state.turnStartDate, add(this.state.turnStartDate, {days: 13}));

		//Remove the newly completed event
		eventsToComplete.splice(eventsToComplete.indexOf(eventid), 1);

  		//Remove all completed event IDs from the array
  		this.state.eventsCompleted.map((completedEvent) => {
  			if(eventsToComplete.includes(completedEvent.eventID)) {
  				eventsToComplete.splice(eventsToComplete.indexOf(completedEvent.eventID), 1);
  			}
  		});

		//Check if the turn needs to be advanced
		let updatedTurnStartDate = this.state.turnStartDate;
		let updatedCurrentSprint = this.state.currentSprint;
  		while(eventsToComplete.length == 0 && !updatedGameEnded){
			//If all events are complete advance the turn counter
			updatedTurnStartDate = add(updatedTurnStartDate, {weeks: 2});

			//Update eventsToComplete to detect turns with no events
			eventsToComplete = this.getEventIDsBetween(updatedTurnStartDate, add(updatedTurnStartDate, {days: 13}));

			//Advance the sprint number
			updatedCurrentSprint += 1;
  		}

		this.setState((state, props) => ({
			pollData: updatedData,
			eventsCompleted: [...state.eventsCompleted, eventCompleted],
			turnStartDate: updatedTurnStartDate,
			currentSprint: updatedCurrentSprint,
			gameEnded: updatedGameEnded
		}));
    };

	checkIfPlayerWon = (eventScore, eventID) =>{
		if(eventID == this.state.lastEventID){
			this.setState({playerScore: eventScore/eventID});
			if(this.state.playerScore > .5){
			  this.setState({hasPlayerWon : true});
			}
			else{
			  this.setState({hasPlayerWon : false});
			}
		}
	}

	//Returns all of the event IDs between 2 dates
	getEventIDsBetween = (turnStartDate, turnEndDate) => {
		let eventsBetween = [];
		Object.values(events).map((event) => {
			let eventDate = new Date(event.year, event.month, event.day, 0, 0, 0, 0);
			if(!(isBefore(eventDate, turnStartDate) || isAfter(eventDate, turnEndDate))) {
        		event.status = 0;
        		eventsBetween.push(event.id);

			}
		});
		return eventsBetween;
	}

	/*checks if the passed in email is already in the list of current emails. If it is not then it returns True, else if it already exists in the list it returns False
	@param  {emails}   The array of the currentEmails displayed.
	@param  {foundEmail}   The email that wants to be added to the current emails.
	*/
	ifExists(emails, foundEmail){
		for(var i in emails) {
			if(emails[i].currentSprint == foundEmail.currentSprint)
			{
				return false;
			}
		}
		return true;
	}


	/*This function gets the current emails needed for the current sprint.
	@param  {emails} The list of emails to be assessed and added to the current email list.
	*/
	setCurrentEmail(emails) {
		for(var i in emails) {
			if(emails[i].currentSprint == this.state.currentSprint)
			{
				if(this.ifExists(this.state.currentEmails, emails[i])){
					this.setState({currentEmails: [...this.state.currentEmails, emails[i]]});
				}
			}
		}
	}


	/*This function allows the calendar to update the turn date which allows the player to progress
	through the game.*/
	updateTurnStartDate  = () => {
		var newdate = addDays(13, this.state.turnStartDate);
		this.setState(
			{
				turnStartDate :  newdate//Moves the turn date up by 2 weeks.
			}
		)
	}

	/**
	 * This function is passed down to Intro so once the video has ended the game can be rendered
	 */
	handleVideoEnd = () => {
		this.setState({renderVideo: false});
	}

	render(){
		return(
			(this.state.renderVideo || this.state.gameEnded) //Check if the intro video or the game should be rendered
			?(//Render video
				<div>
					<img className="desktop" src={desktop} alt="desktop"/>
					<Intro endedCallback={this.handleVideoEnd}/>
				</div>
			)
			:((this.state.gameEnded) //Check if the game has ended
				?((this.state.hasPlayerWon) //Check if the player has won
					?(//Render Good Ending
						<div>
							<img className="desktop" src={desktop} alt="desktop"/>
							<GoodEnding/>
						</div>
					)
					:(//Render Bad Ending
						<div>
							<img className="desktop" src={desktop} alt="desktop"/>
							<BadEnding/>
						</div>
					)
				)
				:(//render game
					//Adding history allows us to start on Email instead of the '/' page
					<Router history={createBrowserHistory().push('./Email')}>
						<div id="screen">
							<audio controls autoPlay loop id="main-music">
								<source src={mainMusicMP3} type="audio/mpeg"></source>
								<source src={mainMusicWAV} type="audio/wav"></source>
								Your Browser does not support the audio element.
							</audio>

							{this.setCurrentEmail(emails)}
							<img className="desktop" src={desktop} alt="desktop"/>
							<nav>
								<Link to='/Calendar'> {/*Button to Calendar*/}
									<Button>
										<span>Calendar</span>
									</Button>
								</Link>

								&nbsp;
								&nbsp; {/*This adds spaces between the buttons*/}
								&nbsp;

								<Link to='/Email'>
									<Button> {/*Button to Email*/}
										<span>Email</span>
									</Button>
								</Link>

								&nbsp;
								&nbsp; {/*This adds spaces between the buttons*/}
								&nbsp;

								<Link to='/Map'>
									<Button> {/*Button to Map*/}
										<span>Map</span>
									</Button>
								</Link>

								&nbsp;
								&nbsp; {/*This adds spaces between the buttons*/}
								&nbsp;

								<Link to= '/Echo'>
									<Button>
										<span>Echo</span>
									</Button>
								</Link>

								&nbsp;
								&nbsp; {/*This adds spaces between the buttons*/}
								&nbsp;

								<Link to= '/Timeline'>
									<Button>
										<span>Timeline</span>
									</Button>
								</Link>
							</nav>

							<Switch>{/*The switch to click between pages.*/}
								<Route path='/Calendar'>
									<CalendarApp  events={Object.values(events)} eventsCompleted={this.state.eventsCompleted} turnStartDate={this.state.turnStartDate}/>
									<Route path='/Calendar/:id' render={(props)=>{
										return <EventPopup callbackFromMain={this.callback} event={events[props.match.params.id]} situation = {Situations[Math.floor(Math.random()* 10)]}/>
										}
									}/>
								</Route>
								<Route path='/Email'>
									<EmailApp emails = {this.state.currentEmails}/>
								</Route>
								<Route path='/Map'>
									<MapApp pollData={this.state.pollData} regionDistrictNames={this.state.regionDistrictNames}/>
									<Route path='/Map/:id' render={(props)=>{
											return <MapRegion region={props.match.params.id} pollData={this.state.pollData} regionDistrictNames={this.state.regionDistrictNames}/>
										}
									}/>
								</Route>
								<Route path='/Echo'>
									<EchoApp echos={echos}/>
								</Route>
								<Route path='/Timeline'>
									<TimelineApp checkIfPlayerWon={this.checkIfPlayerWon} events={Object.values(events)} eventsCompleted={this.state.eventsCompleted}/>
								</Route>
							</Switch>
						</div>
					</Router>
				)
    		)
		);
	}
}

export default MainPage
