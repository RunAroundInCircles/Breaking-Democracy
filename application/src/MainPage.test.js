import React from 'react';
import 'react-router';
import Calendar from './Components/Calendar/Calendar.js';
import MainPage from './MainPage.js';
import renderer from 'react-test-renderer';
import ReactDOM from 'react-dom';
import {
  BrowserRouter as Router,
  Switch,
  Route,
  Link
} from "react-router-dom";

test('renders calendar page', () => {
	const tree = renderer
	.create(<Router><Link page="http://localhost:3000/Calendar">Calendar</Link> </Router>)
	.toJSON();
	expect(tree).toMatchSnapshot();
});

test('renders email page', () => {
	const tree = renderer
	.create(<Router><Link page="http://localhost:3000/Email">Email</Link> </Router>)
	.toJSON();
	expect(tree).toMatchSnapshot();
});

test('renders map page', () => {
	const tree = renderer
	.create(<Router><Link page="http://localhost:3000/Map">Map</Link> </Router>)
	.toJSON();
	expect(tree).toMatchSnapshot();
});

test('renders echo page', () => {
	const tree = renderer
	.create(<Router><Link page="http://localhost:3000/Echo">Echo</Link> </Router>)
	.toJSON();
	expect(tree).toMatchSnapshot();
});

test('renders timeline page', () => {
	const tree = renderer
	.create(<Router><Link page="http://localhost:3000/Timeline">Timeline</Link> </Router>)
	.toJSON();
	expect(tree).toMatchSnapshot();
});