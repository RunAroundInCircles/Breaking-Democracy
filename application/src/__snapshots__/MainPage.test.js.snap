// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders calendar page 1`] = `
Array [
  <a
    href=""
    onClick={[Function]}
    page="http://localhost:3000/Calendar"
  >
    Calendar
  </a>,
  " ",
]
`;

exports[`renders echo page 1`] = `
Array [
  <a
    href=""
    onClick={[Function]}
    page="http://localhost:3000/Echo"
  >
    Echo
  </a>,
  " ",
]
`;

exports[`renders email page 1`] = `
Array [
  <a
    href=""
    onClick={[Function]}
    page="http://localhost:3000/Email"
  >
    Email
  </a>,
  " ",
]
`;

exports[`renders map page 1`] = `
Array [
  <a
    href=""
    onClick={[Function]}
    page="http://localhost:3000/Map"
  >
    Map
  </a>,
  " ",
]
`;

exports[`renders timeline page 1`] = `
Array [
  <a
    href=""
    onClick={[Function]}
    page="http://localhost:3000/Timeline"
  >
    Timeline
  </a>,
  " ",
]
`;
